### PR TITLE
[JENKINS-64657] Create a DoxygenParser based on the existing GccParser

### PR DESCRIPTION
**Pull request by lcmrvk**
Added a new `DoxygenParser` that extends the existing `GccParser` since the warnings messages are somewhat different.
For details refer to [JENKINS-64657](https://issues.jenkins.io/browse/JENKINS-64657).

### CODE REVIEW

**Review by uhafner — COMMENTED** [review]
Can you please create a `DoxygenParser` class in this module that invokes the new constructor here with the pattern of the warnings plugin (`(?:(.+?):(\\d+):(?:(\\d+):)?)? ?([wW]arning|.*[Ee]rror): (.*)$`)? 

Then we see if the tests still work here. And an additional test with the log part that this PR fixes would be nice as well.

**Comment by lcmrvk** [comment]
I've added the log that this PR is supposed to fix and some additional tests.
It's not perfect, but this catches more warnings than before:

On the pro side:
- warning about deprecated tags and other deprecation messages are matched
- warnings with structure such as `Warning: source ../path/to/file is not a readable file or directory... skipping.` are matched as well.

On the cons side:
- in warnings with structure like `Warning: source ../path/to/file is not a readable file or directory... skipping.`, the path leaks to the `message` part and the `path` is reported as `-`.

File: src/main/java/edu/hm/hafner/analysis/parser/Gcc4CompilerParser.java [hunk 1]
```diff
@@ -30,6 +30,14 @@ public Gcc4CompilerParser() {
         super(GCC_WARNING_PATTERN);
     }
 
+    /**
+     * Creates a new instance of {@link Gcc4CompilerParser} with specified pattern.
+     * @param pattern a regex pattern to be used instead of the default one
+     */
+    public Gcc4CompilerParser(final String pattern) {
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/parser/Gcc4CompilerParser.java:37]:
```suggestion
    Gcc4CompilerParser(final String pattern) {
```

File: src/main/java/edu/hm/hafner/analysis/parser/Gcc4CompilerParser.java [hunk 2]
```diff
@@ -30,6 +30,14 @@ public Gcc4CompilerParser() {
...
+        super(pattern == null || pattern.isEmpty() ? GCC_WARNING_PATTERN : ANT_TASK + pattern);
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/parser/Gcc4CompilerParser.java:38]: Fail fast if possible:
```suggestion
        super(pattern);
```

**Review by uhafner — CHANGES_REQUESTED** [review]
Thanks for your PR! Just some minor things.

File: src/main/java/edu/hm/hafner/analysis/parser/Gcc4CompilerParser.java (hunk 2 above)
Comment (lcmrvk) [reply src/main/java/edu/hm/hafner/analysis/parser/Gcc4CompilerParser.java:38]: should this be just `pattern` or `ANT_TASK + pattern`

Comment (uhafner) [reply src/main/java/edu/hm/hafner/analysis/parser/Gcc4CompilerParser.java:38]: I think this should be solely `pattern`. I think that the deriving class should have full control over the pattern.

File: src/main/java/edu/hm/hafner/analysis/parser/DoxygenParser.java [hunk 3]
```diff
@@ -14,7 +14,7 @@
     /**
      * Creates a new instance of {@link DoxygenParser}.
      */
-    public DoxygenParser() {
+    DoxygenParser() {
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/parser/DoxygenParser.java:17]: This one needs to be public, because we call it in the warnings plugin!

**Review by uhafner — CHANGES_REQUESTED** [review]

File: src/main/java/edu/hm/hafner/analysis/parser/DoxygenParser.java (hunk 3 above)
Comment (lcmrvk) [reply src/main/java/edu/hm/hafner/analysis/parser/DoxygenParser.java]: :+1:

File: src/main/java/edu/hm/hafner/analysis/parser/Gcc4CompilerParser.java (hunk 2 above)
Comment (lcmrvk) [reply src/main/java/edu/hm/hafner/analysis/parser/Gcc4CompilerParser.java:38]: OK, I'll add the ANT_TASK in the DoxygenParser class, same as it's Gcc4CompilerParser. Without it there were some tests that failed.

**Review by uhafner — APPROVED** [review]

**Comment by uhafner** [comment]
Thanks! I'll make a new release then we can reference the parser in the corresponding warnings PR.

**Pull request merged**